### PR TITLE
Expose derivation option in AnyAddress, HDWallet

### LINKED ISSUE #2682: Expose versions with derivation, for AnyAddress, HDWalletGetKey
The following methods don't have an exposed version with derivation input:
- `TWAnyAddressCreateWithPublicKey`
- `TWHDWalletGetKeyForCoin`

They should be added for completeness. Implementation exists internally (C++).

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Expose derivation option on `TWAnyAddressCreateWithPublicKey` and `TWHDWalletGetKey`.

New variants:
- `TWAnyAddressCreateWithPublicKeyDerivation` similar to `TWAnyAddressCreateWithPublicKey`
- `TWHDWalletGetAddressDerivation` similar to `TWHDWalletGetAddressForCoin`
- `TWHDWalletGetKeyDerivation` similar to `TWHDWalletGetKeyForCoin`, 

but with an extra `derivation` parameter.

Fixes #2682.

## How to test

Unit tests.

## Types of changes

* New feature (minor, non-breaking change which adds functionality)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: include/TrustWalletCore/TWHDWallet.h [hunk 1]
```diff
@@ -100,7 +100,8 @@ TWData* _Nonnull TWHDWalletEntropy(struct TWHDWallet* _Nonnull wallet);
 TW_EXPORT_METHOD
 struct TWPrivateKey* _Nonnull TWHDWalletGetMasterKey(struct TWHDWallet* _Nonnull wallet, enum TWCurve curve);
 
-/// Generates the default private key for the specified coin.
+/// Generates the default private key for the specified coin, using default derivation.
+/// See also TWHDWalletGetKey, TWHDWalletGetKeyDerivation
```
Comment (miloserdow) [inline include/TrustWalletCore/TWHDWallet.h:104]: Use \see

File: include/TrustWalletCore/TWHDWallet.h [hunk 2]
```diff
@@ -109,15 +110,27 @@ struct TWPrivateKey* _Nonnull TWHDWalletGetMasterKey(struct TWHDWallet* _Nonnull
 TW_EXPORT_METHOD
 struct TWPrivateKey* _Nonnull TWHDWalletGetKeyForCoin(struct TWHDWallet* _Nonnull wallet, enum TWCoinType coin);
 
-/// Generates the default address for the specified coin (without exposing intermediary private key).
+/// Generates the default address for the specified coin (without exposing intermediary private key), default derivation.
+/// See also TWHDWalletGetAddressDerivation
```
Comment (miloserdow) [inline include/TrustWalletCore/TWHDWallet.h:114]: Use \see also

File: include/TrustWalletCore/TWHDWallet.h [hunk 3]
```diff
@@ -109,15 +110,27 @@ struct TWPrivateKey* _Nonnull TWHDWalletGetMasterKey(struct TWHDWallet* _Nonnull
...
 ///
 /// \param wallet non-null TWHDWallet
 /// \param coin  a coin type
 /// \return return the default address for the specified coin as a non-null TWString
 TW_EXPORT_METHOD
 TWString* _Nonnull TWHDWalletGetAddressForCoin(struct TWHDWallet* _Nonnull wallet, enum TWCoinType coin);
 
+/// Generates the default address for the specified coin and derivation (without exposing intermediary private key).
+/// See also TWHDWalletGetAddressForCoin
```
Comment (miloserdow) [inline include/TrustWalletCore/TWHDWallet.h:123]: \see

File: include/TrustWalletCore/TWHDWallet.h [hunk 4]
```diff
@@ -127,6 +140,17 @@ TWString* _Nonnull TWHDWalletGetAddressForCoin(struct TWHDWallet* _Nonnull walle
 TW_EXPORT_METHOD
 struct TWPrivateKey* _Nonnull TWHDWalletGetKey(struct TWHDWallet* _Nonnull wallet, enum TWCoinType coin, TWString* _Nonnull derivationPath);
 
+/// Generates the private key for the specified derivation.
+/// See also TWHDWalletGetKey, TWHDWalletGetKeyForCoin
```
Comment (miloserdow) [inline include/TrustWalletCore/TWHDWallet.h:144]: \see

**Review by Milerius — DISMISSED** [review]
Good job, suggesting to add 1-2 tests in swift/kotlin

**Review by Milerius — APPROVED** [review]

File: android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestAnyAddress.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Test
+
+class TestAnyAddress {
+    init {
+        System.loadLibrary("TrustWalletCore");
+    }
+
+    val ANY_ADDRESS_TEST_ADDRESS = "bc1qcj2vfjec3c3luf9fx9vddnglhh9gawmncmgxhz"
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/utils/TestAnyAddress.kt:25]: Usually just use lower case variable name in Kotlin/Swift classes

File: tests/interface/TWAnyAddressTests.cpp [hunk 6]
```diff
@@ -169,3 +170,34 @@ TEST(AnyAddress, Data) {
...
+
+TEST(TWAnyAddress, createFromPubKey) {
+    constexpr auto pubkey = "02753f5c275e1847ba4d2fd3df36ad00af2e165650b35fe3991e9c9c46f68b12bc";
+    const auto pubkey_twstring = STRING(pubkey);
+    const auto pubkey_data = WRAPD(TWDataCreateWithHexString(pubkey_twstring.get()));
+    const auto pubkey_obj = WRAP(TWPublicKey, TWPublicKeyCreateWithData(pubkey_data.get(), TWPublicKeyTypeSECP256k1));
+
+    const auto addr = WRAP(TWAnyAddress, TWAnyAddressCreateWithPublicKey(pubkey_obj.get(), TWCoinTypeBitcoin));
+
+    EXPECT_EQ(std::string(TWStringUTF8Bytes(WRAPS(TWAnyAddressDescription(addr.get())).get())), "bc1qcj2vfjec3c3luf9fx9vddnglhh9gawmncmgxhz");
```
Comment (hewigovens) [inline tests/interface/TWAnyAddressTests.cpp:182]: I think you can use `assertStringsEqual` to simplify a little bit

**Review by hewigovens — APPROVED** [review]

**Review by hewigovens — APPROVED** [review]

**Review by Milerius — APPROVED** [review]

**Pull request merged**